onnxruntime/core/providers/qnn/builder/opbuilder/matmulnbits_op_builder.cc

```diff
@@ -0,0 +1,351 @@
+  // Float16 Dlc serialization failing, Skipping float16 support for this op builder
+  // TODO :: Add cast from float16 for float32 for scales to enable float16 support.
+  {
+    const NodeUnitIODef& input_tensor = inputs[0];
+    TensorInfo input_info{};
+    ORT_RETURN_IF_ERROR(qnn_model_wrapper.GetTensorInfo(input_tensor, input_info));
+    ORT_RETURN_IF_ERROR(utils::GetQnnDataType(input_tensor.quant_param.has_value(), input_tensor.node_arg.TypeAsProto(), input_datatype));
+    ORT_RETURN_IF(input_datatype != QNN_DATATYPE_FLOAT_32, "Unsupported Input datatype");
+  }
+
+  // 2. weight : weight supported with packed int4 into int8.
+  {
+    const NodeUnitIODef& input_tensor = inputs[1];
+    TensorInfo input_info{};
+    ORT_RETURN_IF_ERROR(qnn_model_wrapper.GetTensorInfo(input_tensor, input_info));
+
+    const std::vector<uint32_t> input_shape = input_info.shape;
+    const int64_t total_elements = std::accumulate(input_shape.begin(), input_shape.end(), 1, std::multiplies<uint32_t>());
+    ORT_RETURN_IF_NOT(((total_elements * 2) == (N * K)), "Invalid B dimensions. Qnn Gpu Only Supports MatMulNBits with bits == 4 in packed format");
+  }
+
+  // 3. scales : scales only float32 datatype
+  {
+    const NodeUnitIODef& input_tensor = inputs[2];
+    TensorInfo input_info{};
+    ORT_RETURN_IF_ERROR(qnn_model_wrapper.GetTensorInfo(input_tensor, input_info));
+    ORT_RETURN_IF_ERROR(utils::GetQnnDataType(input_tensor.quant_param.has_value(), input_tensor.node_arg.TypeAsProto(), input_datatype));
+    ORT_RETURN_IF(input_datatype != QNN_DATATYPE_FLOAT_32, "Unsupported Input datatype");
+  }
+
+  // 4. If input 3 exists, it has to be zero point.
+  if (inputs.size() > 3 && inputs[3].node_arg.Exists()) {
+    const NodeUnitIODef& input_tensor = inputs[3];
+    TensorInfo input_info{};
+    ORT_RETURN_IF_ERROR(qnn_model_wrapper.GetTensorInfo(input_tensor, input_info));
+    ORT_RETURN_IF_ERROR(utils::GetQnnDataType(input_tensor.quant_param.has_value(), input_tensor.node_arg.TypeAsProto(), datatype));
+    ORT_RETURN_IF((datatype != QNN_DATATYPE_UINT_8), "Invalid zero point datatype.");
+
+    std::vector<uint8_t> per_block_uint8_offset;
+    const auto& zero_points_tensor_name = input_tensor.node_arg.Name();
+    const auto& zero_points_tensor_proto = qnn_model_wrapper.GetConstantTensor(zero_points_tensor_name);
+    ORT_RETURN_IF_ERROR(qnn_model_wrapper.UnpackInitializerData(*zero_points_tensor_proto, per_block_uint8_offset));
+
+    ORT_RETURN_IF_NOT((per_block_uint8_offset.size() * 2) == (num_blocks * sizeof(uint8_t)), "Only packed uint4 into uint8 offset supported by op builder");
+    const uint8_t expected_offset_value = 0b10001000;
+    for (int32_t i = 0; i < gsl::narrow_cast<int32_t>(num_blocks / 2); i++) {
+      ORT_RETURN_IF_NOT(per_block_uint8_offset[i] == expected_offset_value, "Unsupported zero point value");
+    }
+  }
+
+  ORT_RETURN_IF(inputs.size() > 4 && inputs[4].node_arg.Exists(), "Unsupported inputs g_idx or bias");
```

Comment:
nit: would have to also check `inputs[5]` to account for "bias" as described in the status message